api/envoy/config/route/v3/route_components.proto

```diff
@@ -818,6 +818,11 @@ message RouteAction {
     // :ref:`traffic splitting <config_http_conn_man_route_table_traffic_splitting_split>`
     // for additional documentation.
     WeightedCluster weighted_clusters = 3;
+
+    // [#not-implemented-hide:]
+    // Requests are handled locally in the xDS client itself, rather than using any cluster. This
+    // is used in xDS clients like the gRPC server that do not forward traffic to other servers.
+    bool is_local_server = 37;
```

Comment:
I guess my opening comment would be: couldn't you just use a synthetic "local_server" cluster? I'm trying to to understand the utility of having this in the API?

Reply:
The main benefit is that it avoids the need to use a CDS resource that we don't actually need or care about, which is both easier for users (they don't need to configure the resource) and reduces network usage and startup time.

I think it also allows us to more accurately reflect where the config error is when the xDS server sends the wrong configuration.  If the gRPC server gets a RouteConfiguration that points to a Cluster that is *not* the local server, the gRPC server will have to reject the config, but since it doesn't know there's an error until it sees the Cluster resource, that's where it would need to NACK.  But it doesn't really make sense to NACK the Cluster resource, because that resource may be perfectly valid even though it can't be used by the gRPC server.  In contrast, if the RouteConfiguration itself tells the gRPC server whether the destination is local or not, then the gRPC can immediately NACK the RouteConfiguration, because it can know immediately that the RouteConfiguration is telling it something invalid.

Reply:
Sorry what I was suggesting is to just set "cluster" (above) to "local_server" or whatever. You don't need a CDS record.

Reply:
That's certainly doable, but it seems a little special-casey.  That implies that "local_cluster" becomes some magic sentinel value that clients will need to know to handle specially.  What if someone actually creates a cluster with that name?

Another option would be to simply have the gRPC server ignore the cluster field completely, so any value the server puts there would be acceptable.  But that also seems a little strange from the perspective of API semantics; if the xDS server is expecting the xDS client to actually use this value and it just gets ignored, that could lead to unexpected behavior.

Yet another option might be to have a new cluster type that indicates a local server.  That would still require us to fetch a CDS resource that we don't really need, but it would at least be semantically clear what was intended.  And when we eventually get around to supporting on-demand clusters, maybe we could provide an option in the RouteAction to just inline the cluster resource itself?

Reply:
> That's certainly doable, but it seems a little special-casey. That implies that "local_cluster" becomes some magic sentinel value that clients will need to know to handle specially. What if someone actually creates a cluster with that name?

Since the gRPC server is going to effectively ignore it anyway does it matter?

> Another option would be to simply have the gRPC server ignore the cluster field completely, so any value the server puts there would be acceptable. But that also seems a little strange from the perspective of API semantics; if the xDS server is expecting the xDS client to actually use this value and it just gets ignored, that could lead to unexpected behavior.

Per my comment above this seems OK to me.

> Yet another option might be to have a new cluster type that indicates a local server. That would still require us to fetch a CDS resource that we don't really need, but it would at least be semantically clear what was intended. And when we eventually get around to supporting on-demand clusters, maybe we could provide an option in the RouteAction to just inline the cluster resource itself?

Yeah this could work, but seems overkill.

Ultimately I don't feel very strongly about it, I'm mostly just trying to avoid API creep if we can avoid it. Any other @envoyproxy/api-shepherds have an opinion?

Reply:
So, I think this comes up in other places, when we have Envoy as an origin server and we're never sending to a backend, e.g. some filter chain which implements content streaming or the like. In this case, seems like it would be good to say something like "there's no backing service", which is a little different to "local server". Is there a way to express a field that can serve both functions?

Reply:
We already have DirectResponse which is roughly this. Could we somehow make DirectResponse handle all of these cases somehow?

Reply:
DirectResponse makes sense for when router itself is serving, but in other cases it might happen earlier.

Reply:
Right, I guess my point is that this seems like another case of local serving. What if we have a web-server filter that serves from disk? How would we model that? Per-filter matching config along with a "no backing service" route?

Reply:
If we add a specific sentinel for "no backing service" personally I would prefer "terminal" but I don't feel that strongly about it, and I think we would need to decide what the router should do if it sees this. Probably return 503 as a misconfig?

Reply:
Sorry for taking so long to reply to this; we got pulled off on other things, but we're now coming back to this.

My one concern about using `DirectResponseAction` (or some other new type of action in the [`Route.action` oneof](https://github.com/envoyproxy/envoy/blob/0c3702839c2f733654f1e74ec7708ffb67c6ec74/api/envoy/config/route/v3/route_components.proto#L219)) instead of `RouteAction` is that there may be fields in `RouteAction` that we might wind up needing on the server side.  @ejona86 and @dfawley and I took a cursory look over the fields of `RouteAction`, and the existing fields that we could imagine using at some point on the gRPC server would be these:
- max_stream_duration
- rate_limits
- cors

And there could be other fields added to `RouteAction` in the future that might also be useful on the gRPC server side.  So if we go with a different type of action, then we'll need to commit to the possibility of copying fields from `RouteAction` to that new action type whenever they're needed in the gRPC server, which seems like it could ultimately result in a fair amount of duplication.  If we're okay with that duplication, then that approach would certainly work, but it seems to me that we could avoid it by sticking with `RouteAction`.

I'm open to calling this field whatever seems best.  I'd be fine with something like "no_backing_service" or "no_cluster" or some such thing.

In terms of how Envoy responds to this config, yes, I think the router filter could return a 503.  Alternatively, if you know that there aren't any filters that will ever generate their own responses, you could even have Envoy NACK a config with this field set at validation time.

How should we proceed with this?

Reply:
Ideally we would migrate things that common to either `typed_per_filter_config` when they are per-route and filter related (e.g. CORS, rate limits) and have common stream related properties in a refactored message.

However, I'm not sure how practical that is unless someone wants to drive it. How much do you need  today?

I think I'm good with `no_cluster` or `terminal` here naming wise.

Reply:
We don't need any of those fields immediately, but it's possible that we will in the future.

I think I buy the argument that things like CORS and rate limits should ideally be moved out of `RouteAction` and into `typed_per_filter_config` for the respective filters.  I'd imagine that max_stream_duration would still stay in RouteAction.  But if that's the only exception (or if it's indicative of the fact that the exceptions will be a minority), then it may be feasible to simply duplicate the fields for the exceptions in the new action type.

So this would mean that if we go with a different type of action now and then we later decide that we need some of this functionality, we will then need to do the work to move the needed fields from `RouteAction` into `typed_per_filter_config` (or, in the exception cases, duplicate the fields from `RouteAction` into the new action type).  I think we can live with that -- it's more work when the time comes, but it will yield a cleaner outcome.

So I think we're okay with either of these options:

1. Adding a new field here called `no_cluster`.
2. Introducing a new type of action for terminal servers.

Which would you prefer?

Reply:
I personally prefer (2) since most `RouteAction` fields are irrelevant to origin serving. I'd call it `OriginAction` and be done. It's definitely not `DirectResponse` in your case.

Reply:
I agree that `typed_per_filter_config` seems to make more sense for cors/rate_limits. Given the large number of fields, it is telling that only max_stream_duration remains.

I like the `OriginAction` name and approach. The vast majority of fields in `RouteAction` have no way they could apply. It'll be annoying to deal with RateLimit/MaxStreamDuration if/when we need them, but nothing more. CORS shouldn't impact us for the foreseeable future. Technically DirectResponse would be an "origin" action as well, but I doubt we'll find any words that obviously differentiate those two actions.

Reply:
Okay, we'll go with approach (2).

I find the name `OriginAction` a little confusing, because this isn't the origin of the request; it's the destination.  How about `NonForwardingAction`?

Anyway, it looks like this PR can't be reopened, because it was created against `master`, which has subsequently been renamed `main`.  So I've created #16144, and we can bikeshed the naming there.

Reply:
[origin server](https://tools.ietf.org/html/rfc2616#page-10) has been long-standing HTTP-speak for the HTTP server that is the final destination of a request. In fuller HTTP-speak, it is the server that produces a resource (where resource implies response); i.e., not a proxy. "Origin" these days is a sort of similar concept to "authority" and is used in a lot of web discussions. CORS stands for "cross-origin resource sharing". Browsers implement the "same-origin policy." All of these usages of "origin" are talking about _resources_ and the servers producing them.

I guess `OriginServerAction` is more precise, but dunno if it matters.

Reply:
Ah, interesting -- I wasn't actually aware that that was a widely established term.

I've already sent #16144 using the name `NonForwardingAction`, but feel free to comment on that PR if you'd prefer changing it to `OriginServerAction`.